documents/write_acls.go

```diff
@@ -84,3 +86,121 @@ func newChangedField(p proofs.Property, leaf *proofs.LeafNode, old bool) changed
 	cf.new = v
 	return cf
 }
+
+// transitionRulesFor returns a copy all the transition rules for the account.
+func (cd *CoreDocument) transitionRulesFor(account identity.DID) (rules []coredocumentpb.TransitionRule) {
+	for _, rule := range cd.Document.TransitionRules {
+		for _, rk := range rule.Roles {
+			role, err := getRole(rk, cd.Document.Roles)
+			if err != nil {
+				continue
+			}
+
+			if _, ok := isAccountInRole(role, account); !ok {
+				continue
+			}
+
+			rules = append(rules, coredocumentpb.TransitionRule{
+				RuleKey:   copyBytes(rule.RuleKey),
+				Roles:     copyByteSlice(rule.Roles),
+				MatchType: rule.MatchType,
+				Field:     copyBytes(rule.Field),
+				Action:    rule.Action,
+			})
+		}
+	}
+
+	return rules
+}
+
+func copyBytes(data []byte) []byte {
+	if data == nil {
+		return nil
+	}
+
+	nb := make([]byte, len(data), len(data))
+	copy(nb, data)
+	return nb
+}
+
+func copyByteSlice(data [][]byte) [][]byte {
+	nbs := make([][]byte, len(data), len(data))
+	for i, b := range data {
+		nbs[i] = copyBytes(b)
+	}
+
+	return nbs
+}
+
+// validateTransitions validates the changedFields based on the rules provided.
+// returns an error if any changedField violates the rules.
+func validateTransitions(rules []coredocumentpb.TransitionRule, changedFields []changedField) error {
+	cfMap := make(map[string]struct{})
+	for _, cf := range changedFields {
+		cfMap[cf.name] = struct{}{}
+	}
+
+	for _, rule := range rules {
+		for _, cf := range changedFields {
+			if isValidTransition(rule, cf) {
+				delete(cfMap, cf.name)
+			}
+		}
+	}
+
+	if len(cfMap) == 0 {
+		return nil
+	}
+
+	var err error
+	for k := range cfMap {
+		err = errors.AppendError(err, errors.New("invalid transition: %s", k))
+	}
+
+	return err
+}
+
+func isValidTransition(rule coredocumentpb.TransitionRule, cf changedField) bool {
+	// changed property length should be at least equal to rule property
+	if len(cf.property) < len(rule.Field) {
+		return false
+	}
+
+	// if the match type is prefix, get the compact property till prefix
+	v := cf.property
+	if rule.MatchType == coredocumentpb.FieldMatchType_FIELD_MATCH_TYPE_PREFIX {
+		v = v[:len(rule.Field)]
+	}
+
+	// check the properties are equal
+	if !bytes.Equal(v, rule.Field) {
+		return false
+	}
+
+	// check if the action is allowed
+	// for now, we have only edit action
+	// edit allows following
+	// 1. update: editing a value
+	// 2. set: setting a new value ex: adding to slice or map
+	// 3. delete: deleting the new value ex: removing from slice or map
+	// Once we have more actions, like set, increment etc.. we can do those checks here
+	return true
+}
+
+// AccountCanUpdate validates the changes made by the account in the new document.
+// returns error if the transitions are not allowed for the account
+func (cd *CoreDocument) AccountCanUpdate(ncd *CoreDocument, account identity.DID, docType string) error {
```

Comment:
Same here with `IDCanUpdate`

Reply:
I'm leaning towards `CollaboratorCanUpdate` since ID seems very generic. What do you think ?

Reply:
Yep, collaborator fits well too.

Reply:
I think `AccountCanUpdate` makes sense as ID is just a single part of an account. Also i don't mind `CollaboratorCanUpdate`.